### PR TITLE
Add Qucik Surf extension

### PR DESCRIPTION
**Pull request by koinzhang**
This extension is for users who Surf frequently between browsers or mailbox applications, choose your board and surf fast!

## Description

**Main features:**

1. Specify a browser to open a ***URL*** or ***Text*** that is selected or on the clipboard.
2. Specify the mailbox application to edit email with ***Email addresses*** or ***mailto:*** link (in the form of "mailto:hello@raycast.com") that is selected or on the clipboard.
3. Support Search Bar input:
   Enter ***URL*** or ***Text*** to search or open web page quickly.
   Enter ***Email addresses*** or ***mailto:***  followed by email address to edit email (Separate multiple addresses with " ***;*** ").
4. Recommend to set shortcut keys for this extension in order to Quick Quick Surf.

**Preference:**
*Surf Engine* - the default search engine for searching text.
*Sort by* - Surfboard sorting method.

**Known issues:**
If the application in Surfboard is uninstalled, it will not be automatically removed from Surfboard, please press the "cmd+enter" shortcut on the "More Boards"screen to remove the invalid application.

## Screencast

First, "More Boards"，add application you want to use.
More Boards-1
More Boards-2

Quick search ***Text*** that is selected, on the clipboard or from input using a specified browser
Search Text

Quick open ***URL*** that is selected, on the clipboard or from input using a specified browser
Open URL

Quick open ***"mailto"*** link that is selected, on the clipboard or from input using a specified browser
mailto

## 

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed in the `media` folder

### CODE REVIEW

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @CheatGZ 👋 

Thanks for your contribution, it seems like a very useful idea that you're building here.

I have some feedback for you to address:

If clipboard and nothing is selected, should there then be the `Surfboard` section? Seems like if I use `Mimestream` with empty clipboard, it does something weird 😄 

66rmOmQfN23lSz6W

FIqo3CM93EO5a74H

- When opening a URL from clipboard, some elements in surfboard switches place, is that as intended?

https://user-images.githubusercontent.com/38033107/157316461-48e78a32-fe8e-447d-a623-db5c5f14170d.mp4

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by CheatGZ** [comment]
Hi, @pernielsentikaer 👋
Thank you for your review, I have noticed the issues you mentioned. Here is my response.

**The first question is divided into two parts：**
Part I - Empty Clipboard

- The clipboard is not empty, but there is a delay in the appearance of text, because of runAppleScript ("the clipboard"), it will be the clipboard image or file parsed as text, the second screenshot to see the "Type" section of the appearance of the Text.
  Delay reason: Quick Surf detects the selected text first and then detects the clipboard. So it detects the clipboard, the speed will be slower. Therefore, it is more recommended to use the "Select-Surf" method(Select text or URL and open Quick Surf).
- Fix solution:
  When the detection is empty, keep the "Surfboard" section, and use toast to prompt the user to enter text.
  In rare cases, the detection will be empty, and hiding the "Surfboard" Section will make people confused: where is the Surfboard that was added.

<details>
<summary>Clipboard detect delay</summary>

https://user-images.githubusercontent.com/36128970/157384041-81a5ed86-6f1e-4c1f-8415-20114b6f59d5.mp4

</details>


Part II - Mimestream No Respond

- Quick Surf uses the “Open()” provided by Raycast to open links in specific applications, it allows any application to open a URL or "mailto:" link, but if open the URL with a non-browser application, there will be no response and no error will be reported. Similarly, if a non-email client is used to open a "mailto:" link, no response or error will be reported (except for Safari).
- Fix solution:
  There is no better solution now because users can choose any application to open the link, but I don't know how the app will react; Only expect users to open the URL or the "mailto:" link with the appropriate application(browsers or email clients).

**The Second question:**

- I'm glad you noticed this, I did intend to do this. The types of text that the extension can receive are URL, Text, and Email address("mailto:"), so I designed a simple recommendation algorithm that will record the user's habits and sort the "Surfboard" section according to input, for example, if you often use Safari to search for text, Safari will rise when input is text; if you often use Mimestream to edit emails, Mimestream will rise when input is an email address.

<details>
<summary>InputType</summary>

image

https://user-images.githubusercontent.com/36128970/157388750-f5abc024-c506-4d6e-8c82-d9d07c13a531.mp4

</details>


- If users don't like the automatic sorting of the "Surfboard" section, they can choose to sort by name in Preferences.

File: extensions/quick-surf/package.json [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  "name": "quick-surf",
+  "title": "Quick Surf",
+  "description": "Quick Surf between browsers or email clients.",
+  "icon": "quick-surf.png",
+  "author": "koinzhang",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "quick-surf",
+      "title": "Surf with...",
```
Comment (pernielsentikaer) [inline extensions/quick-surf/package.json:12]:
```suggestion
      "title": "Quick Surf",
```
Maybe change title to this?

File: extensions/quick-surf/src/quick-surf.tsx [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+  showToast,
+  Toast,
+  useNavigation,
+} from "@raycast/api";
+import { useEffect, useState } from "react";
+import {
+  assembleItemSource,
+  assembleItemType,
+  fetchDetectedItem,
+  isEmpty,
```
Comment (pernielsentikaer) [inline extensions/quick-surf/src/quick-surf.tsx:19]:
```suggestion
```
Unused import specifier

File: extensions/quick-surf/src/quick-surf.tsx [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+  return (
+    <List.Item
+      id="MoreBoards"
+      title={"More Boards"}
+      icon={{ source: { light: "quick-surf.png", dark: "quick-surf@dark.png" } }}
+      // accessoryTitle={"❯"}
+      accessoryIcon={Icon.ArrowRight}
+      actions={
+        <ActionPanel title="Game controls">
```
Comment (pernielsentikaer) [inline extensions/quick-surf/src/quick-surf.tsx:136]: Game controls? Remember to `Title Case` the corrected

File: extensions/quick-surf/src/quick-surf.tsx [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+      case ItemType.EMAIL:
+        await showHUD("Send Email with " + app.name);
+        break;
+    }
+    await openSurfboard(searchEngineURLBuilder(inputText), app.path);
+    await popToRoot({ clearSearchBar: true });
+  } else {
+    const selectedItem = await fetchDetectedItem();
+    if (selectedItem.type == ItemType.NULL) {
+      await showToast(Toast.Style.Failure, "Nothing, Enter Now!");
```
Comment (pernielsentikaer) [inline extensions/quick-surf/src/quick-surf.tsx:242]: Could be a better error message

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @CheatGZ 

Seems to do better with the things I've tested 🔩 

I have some feedback for you:

- Is it on purpose that the @dark icons have the dark background? Seems like it should be reverted (could also just be me)

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by CheatGZ** [comment]
Hi, @pernielsentikaer.
Thank you for your review. Your comments were very informative and I have made changes accordingly.
Update:
1. Light icons and dark icons are swapped, now the icons are more eye-catching.
2. Command name change to Quick Surf.
3. New detect tip: "Nothing is detected from Selected or Clipboard! "
4. Application of “Suggest” section add tips.
5. Support IP address, support URL without "https://".
6. More action.
7. Redundant code cleaning.

**Review by pernielsentikaer — APPROVED** [review]
Hi @CheatGZ 👋 

Thanks for your update, tested it again, and it does as expected, awesome! 

Approved, lets send it to final review!

**Pull request merged**